frontend/apps/reader/modules/readercoptlistener.lua

```diff
@@ -54,6 +60,42 @@ function ReaderCoptListener:setAndSave(setting, property, value)
     UIManager:broadcastEvent(Event:new("SetStatusLine", self.document.configurable.status_line, true))
 end
 
+function ReaderCoptListener:updateHeader()
+    if self.view.view_mode == "page" then
+        self.ui.rolling:updateBatteryState()
+        self.ui.document:resetBufferCache()
+        UIManager:setDirty(self.view.dialog, "ui",
+            Geom:new{ w = Device.screen:getWidth(), h = self.ui.document:getHeaderHeight()})
+    end
+end
+
+function ReaderCoptListener:setupHeaderRefresh()
+    self.headerRefresh = function()
+        -- Only actually repaint the header only if it's actually visible
+        if self.document.configurable.status_line == 0 then -- is top bar enabled
+            local new_battery_level = Device:getPowerDevice():getCapacity()
```

Comment:
No more need for `UIManager:getTopWidget() == "ReaderUI" ` ?!
The logic still seems not right.
I guess we want:
- in updateHeaderRefreshSchedule(): what we know at this point and can be toggled from the ReaderCoptListener menu items: if they say stuff needs to be updated: schedule - else, don't.
- in self.headerRefresh, which happens 10-50s later: we should check stuff that may have changed in-between from elsewhere (configurable.status_line , self.view.view_mode == "page" ) and can possible avoid doing it and avoid rescheduling.
- in ReaderCoptListener:updateHeader(), possibly the very local stuff like topwidget!=ReaderUI that doesn't prevent a rescheduling.

Dunno if we should try to catch with a onEvent any change in configurable.status_line , self.view.view_mode so we can unschedule everything if header is off OR view_mode="scroll" - and reschedule when header is on AND view_mode="page".

Reply:
@poire-z: I am loosing my joy in contributing with PRs. This is nothing personal. You and the others are polite, helpfull ... professional ...

Please take over this PR and the other one with the bottom bar, if you think they are worthfull enough.

I will stay with KOReader. You will get issues, FRs and so from me, but I won't send no code anymore.

I wish you all the best .
Cheers

Reply:
> I am loosing my joy in contributing with PRs

I'm sorry to hear that.
You actually got quite quickly and well into the schedudling stuff, and you were almost done.
But if I can think of some cases or combinations where it may not be fine, I'm sorry but I just have to mention them - so we can fix them when it's fresh. This is "core" stuff, we don't really want to let it unfinished (unlike "plugins" where we can care a bit less).
And these additional finishing touches is what makes the whole codebase (hopefully) coherent - and make it (hopefully) easier for people like you to get into quickly (unlike crengine).

And yes, it's not always a joy - including for me. Many times have I spent a few days or weeks making a PR, considered it perfect, pushed it - and got reviews about what I would consider nitpicks, that needed a few more hours of not fun work and the need to re-test the whole thing... (and yes, I sometimes did resent them as harassment in my early PRs :/ so, sorry if it feels that way, but that's not my intention - but I know the feeling).
And it's also not always a joy for me to work on or review bits that I don't care much about, like the top status line or the dual page mode :) which delays my fun work (a bit busy with work's work these last days to work on two columns). But I do it so it doesn't get in the way, and I and other people can move on.
I was actually quite happy you made this PR after I mentionned there was no need to do any crengine tweaks, and a setDirty/painto would probably be enough - as I thought that would have been another thing on my not-fun-but-todo list... And happy too that you were appropriating ReaderCoptListener and pushing new ideas in it.

So, well, if you reconsider, good - if not, I'll probably take over this PR as it's really good and there's not much left to do.